src/loaders/ncbi_downloader/gtdb.py

```diff
@@ -0,0 +1,311 @@
+        "--download_file_ext",
+        required=True,
+        type=str,
+        nargs="+",
+        help="Download only files that match given extensions",
+    )
+    required.add_argument(
+        "--release_ver",
+        required=True,
+        type=str,
+        choices=gtdb_release_vers,
+        help="GTDB release version",
+    )
+
+    # Optional argument
+    optional.add_argument(
+        f"--{loader_common_names.KBASE_COLLECTION_ARG_NAME}",
+        type=str,
+        default="GTDB",
+        choices=COLLECTIONS,
+        help="Create a collection and link in data to that collection from the overall source data dir",
+    )
+    optional.add_argument(
+        "--root_dir",
+        type=str,
+        default=loader_common_names.ROOT_DIR,
+        help="Root directory",
+    )
+    optional.add_argument(
+        "--source", type=str, default="NCBI", choices=SOURCE, help="Source of data"
+    )
+    optional.add_argument("--threads", type=int, help="Number of threads")
+    optional.add_argument(
+        "--overwrite", action="store_true", help="Overwrite existing files"
+    )
+    optional.add_argument(
+        "--exclude_name_substring",
+        type=str,
+        nargs="+",
+        default=loader_common_names.STANDARD_FILE_EXCLUDE_SUBSTRINGS,
+        help="Files with a specific substring in their names that should be excluded from the download",
+    )
+
+    args = parser.parse_args()
+
+    download_file_ext = args.download_file_ext
+    release_ver = args.release_ver
+    root_dir = args.root_dir
+    source = args.source
+    threads = args.threads
+    overwrite = args.overwrite
+    exclude_name_substring = args.exclude_name_substring
+    kbase_collection = getattr(args, loader_common_names.KBASE_COLLECTION_ARG_NAME)
+
+    if threads and (threads < 1 or threads > cpu_count()):
+        parser.error(f"minimum thread is 1 and maximum thread is {cpu_count()}")
+
+    work_dir = _make_work_dir(
+        root_dir, loader_common_names.SOURCE_DATA_DIR, source, loader_common_names.DEFAULT_ENV
+    )
+    csd = loader_helper.make_collection_source_dir(
+        root_dir,
+        loader_common_names.COLLECTION_SOURCE_DIR,
+        loader_common_names.DEFAULT_ENV,
+        kbase_collection,
+        release_ver,
+    )
+    genome_ids_unprocssed, taxonomy_urls = _fetch_gtdb_genome_ids(release_ver, work_dir)
+    taxonomy_files = [os.path.basename(url) for url in taxonomy_urls]
+    genome_ids = _process_genome_ids(
+        work_dir,
+        genome_ids_unprocssed,
+        download_file_ext,
+        exclude_name_substring,
+        overwrite,
+    )
+    if not genome_ids:
+        print(f"All {len(genome_ids_unprocssed)} genomes files already exist in {work_dir}")
+        loader_helper.create_softlinks_in_csd(csd, work_dir, genome_ids_unprocssed, taxonomy_files)
+        return
+
+    if not threads:
+        threads = max(int(cpu_count() * min(SYSTEM_UTILIZATION, 1)), 1)
+    threads = max(1, threads)
```

Comment:
We should move all the threading stuff into the ncbi downloader helper so it can be shared. No need to reinvent the the multiprocessing wheel every time we want to call those functions

Reply:
Updated!

Reply:
The threading stuff is still all in gtdb.py?

Reply:
https://github.com/kbase/collections/blob/a4c61eaa0b1bb6d0caba947564ec3cd3b9f939cd/src/loaders/ncbi_downloader/gtdb.py#L262

Reply:
I'm talking about the entire threading mechanism, not just calculating the thread count. E.g. if I call the ncbi helper from another script, I just pass it the system utilization and/ or thread count I want and it does the rest

Reply:
Otherwise we have to repeat the threading code for every script that calls the ncbi helper

Reply:
Updated! Moved into ncbi helper.

Reply:
:+1: